// @kakao/eslint-config-styleguide/es3/rules/max-len.js
const {LINT_CATEGORY, AIRBNB_CATEGORY} = require('../../lib/generator/constants');

const title = 'word-wrap은 120자로 유지한다.';

const rule = 'max-len';

const desc = `단, 아래 항목들은 길이 제한에서 제외한다.
  - 주석
  - 한 줄의 끝에 붙이는 주석
  - URL
  - 문자열
  - 정규표현식`;

const options = [
  'error',
  {
    code: 120,
    ignoreComments: true,
    ignoreTrailingComments: true,
    ignoreUrls: true,
    ignoreStrings: true,
    ignoreTemplateLiterals: true,
    ignoreRegExpLiterals: true
  }
];

const cause = null;

const lintCategory = LINT_CATEGORY.STYLISTIC_ISSUES;

const airbnbCategory = AIRBNB_CATEGORY.WHITESPACE;

/* eslint-disable no-undef, no-unused-vars, no-redeclare, no-magic-numbers, no-unused-expressions, no-self-compare, no-cond-assign, default-case, no-unreachable */
/* eslint-disable max-len */
const bad = () => {
  var foo = {bar: 'This is a bar.', baz: {qux: 'This is a qux'}, difficult: 'to read', str1: 'something long', str2: 'long 2'};
};
/* eslint-enable max-len */

const good = () => {
  var foo = {
    bar: 'This is a bar.',
    baz: {qux: 'This is a qux'},
    difficult: 'to read',
    str1: 'something long',
    str2: 'long 2'
  };
};

const best = null;
/* eslint-enable no-undef, no-unused-vars, no-redeclare, no-magic-numbers, no-unused-expressions, no-self-compare, no-cond-assign, default-case, no-unreachable */

module.exports = {
  title,
  rule,
  options,
  desc,
  cause,
  lintCategory,
  airbnbCategory,
  bad,
  good,
  best
};
