// @kakao/eslint-config-styleguide/es2015/.eslintrc.js
module.exports = {
  plugins: ['import'],
  parser: 'babel-eslint',
  rules: {
    'no-const-assign': 'error',
    'no-shadow': 'error',
    'no-var': 'error',
    'prefer-const': 'error',
    'comma-dangle': [
      'error',
      {
        arrays: 'always-multiline',
        objects: 'always-multiline',
        imports: 'never',
        exports: 'never',
        functions: 'never',
      },
    ],
    'no-extra-semi': 'error',
    'prefer-destructuring': [
      'error',
      {
        object: true,
        array: true,
      },
    ],
    'arrow-body-style': [
      'error',
      'as-needed',
    ],
    'arrow-parens': [
      'error',
      'as-needed',
    ],
    'arrow-spacing': [
      'error',
      {
        before: true,
        after: true,
      },
    ],
    'implicit-arrow-linebreak': [
      'error',
      'beside',
    ],
    'prefer-template': 'error',
    'template-curly-spacing': [
      'error',
      'never',
    ],
    'object-shorthand': 'error',
    'generator-star-spacing': [
      'error',
      {
        before: false,
        after: true,
      },
    ],
    'yield-star-spacing': [
      'error',
      {
        before: false,
        after: true,
      },
    ],
    'constructor-super': 'error',
    'no-class-assign': 'error',
    'no-dupe-class-members': 'error',
    'no-this-before-super': 'error',
    'no-empty-pattern': 'error',
    'no-new-symbol': 'error',
    'no-case-declarations': 'error',
    'no-useless-computed-key': 'error',
    'no-return-await': 'error',
    'prefer-rest-params': 'error',
    'prefer-spread': 'error',
    'no-confusing-arrow': [
      'error',
      {
        allowParens: true,
      },
    ],
    'prefer-arrow-callback': 'error',
    'import/first': 'error',
    'import/no-mutable-exports': 'error',
    'no-duplicate-imports': 'error',
    'no-useless-rename': 'error',
  },
};
