// @kakao/eslint-config-styleguide/es2015/rules/comma-dangle.js
const {LINT_CATEGORY, AIRBNB_CATEGORY} = require('../../lib/generator/constants');

const title = '객체/배열인 경우 trailing comma를 사용한다.';

const rule = 'comma-dangle';

const desc = '배열의 경우 여러 줄로 표현될 때만 trailing comma를 사용한다.';

const options = ['error', {
  arrays: 'always-multiline',
  objects: 'always-multiline',
  imports: 'never',
  exports: 'never',
  functions: 'never',
}];

const cause = null;

const lintCategory = LINT_CATEGORY.STYLISTIC_ISSUES;

const airbnbCategory = AIRBNB_CATEGORY.COMMAS;

/* eslint-disable no-unused-vars, no-undef, no-unused-expressions, no-redeclare */
/* eslint-disable comma-dangle */
const bad = () => {
  const obj = {
    bar: 'baz',
    qux: 'quux'
  };

  const arr1 = [1, 2, ];
  const arr2 = [1,
    2, ];
  const arr3 = [
    1,
    2
  ];
};
/* eslint-enable comma-dangle */

const good = () => {
  const obj = {
    bar: 'baz',
    qux: 'quux',
  };

  const arr1 = [1, 2];
  const arr2 = [1,
    2];
  const arr3 = [
    1,
    2,
  ];
};

const best = null;
/* eslint-enable no-unused-vars */
/* eslint-enable no-unused-vars, no-undef, no-unused-expressions, no-redeclare */

module.exports = {
  title,
  rule,
  options,
  desc,
  cause,
  lintCategory,
  airbnbCategory,
  bad,
  good,
  best,
};
