// @kakao/eslint-config-styleguide/es2015/rules/no-empty-pattern.js
const {
  LINT_CATEGORY,
  AIRBNB_CATEGORY,
} = require('../../lib/generator/constants');

const title = '비어있는 비구조화 패턴은 사용하지 않는다.';

const rule = 'no-empty-pattern';

const desc = null;

const options = 'error';

const cause = null;

const lintCategory = LINT_CATEGORY.POSSIBLE_ERRORS;

const airbnbCategory = AIRBNB_CATEGORY.DESTRUCTURING;

/* eslint-disable no-unused-vars, no-undef, no-unused-expressions, no-redeclare */
/* eslint-disable no-empty-pattern */
const foo = {};

const bad = () => {
  const {} = foo;
  const [] = foo;
  const {
    a: {},
  } = foo;

  const {
    a: [],
  } = foo;

  function bad1 ({}) {}

  function bad2 ([]) {}

  function bad3 ({a: {}}) {}

  function bad4 ({a: []}) {}
};
/* eslint-enable no-empty-pattern */

const good = () => {
  const {a = {}} = foo;
  const {b = []} = foo;

  function good1 ({c = {}}) {}

  function good2 ({d = []}) {}
};
/* eslint-disable no-unused-vars, no-undef, no-unused-expressions, no-redeclare */

const best = null;

module.exports = {
  title,
  rule,
  options,
  desc,
  cause,
  lintCategory,
  airbnbCategory,
  bad,
  good,
  best,
};
