// @kakao/eslint-config-styleguide/lib/generator/constants.js
const LINT_CATEGORY = {
  VARIABLES: 'Variables',
  STYLISTIC_ISSUES: 'Stylistic Issues',
  POSSIBLE_ERRORS: 'Possible Errors',
  BEST_PRACTICES: 'Best Practices',
};

const AIRBNB_CATEGORY = {
  TYPES: 'Types',
  TYPE_CASTING_N_COERCION: 'Type casting & coercion',
  VARIABLES: 'Variables',
  NAMING_CONVENTIONS: 'Naming conventions',
  WHITESPACE: 'Whitespace',
  BLOCKS: 'Blocks',
  COMMAS: 'Commas',
  SEMICOLONS: 'Semicolons',
  OBJECTS: 'Objects',
  PROPERTIES: 'Properties',
  FUNCTIONS: 'Functions',
  ARROW_FUNCTIONS: 'Arrow Functions',
  STRINGS: 'Strings',
  ARRAYS: 'Arrays',
  REGEXP: 'RegExp',
  CLASSES_N_CONSTRUCTORS: 'Classes & Constructors',
  ITERATORS_N_GENERATORS: 'Iterators & Generators',
  CONTROL_STATEMENTS: 'Control statements',
  COMPARISON_OPERATORS_N_EQUALITY: 'Comparison Operators & Equality',
  DESTRUCTURING: 'Destructuring',
  MODULES: 'Modules',
  ERRORS: 'Errors',
  COMMENTS: 'Comments',
  ETC: 'Etc',
};

const RULE_DEF_TYPE = {
  SEPARATED_BY_FILE: 'SEPARATED_BY_FILE',
  SEPARATED_BY_FOLDER: 'SEPARATED_BY_FOLDER',
};

const ENV_FLAGS = {
  USE_LOCAL_LINT_RULE: 'USE_LOCAL_LINT_RULE'
};

module.exports = {
  LINT_CATEGORY,
  AIRBNB_CATEGORY,
  RULE_DEF_TYPE,
  ENV_FLAGS,
};
